src/paradigma/tremor/tremor_analysis.py

```diff
@@ -101,4 +106,99 @@ def detect_tremor_io(input_path: Union[str, Path], output_path: Union[str, Path]
 
-    write_df_data(metadata_time, metadata_samples, output_path, 'tremor_meta.json', df)
\ No newline at end of file
+    write_df_data(metadata_time, metadata_samples, output_path, 'tremor_meta.json', df)
+
+
+def aggregate_tremor_power(tremor_power: pd.Series, config: TremorQuantificationConfig):
+
+    tremor_power = np.log10(tremor_power+1) # convert to log scale
+    
+    # calculate median and 90th percentile of tremor power
+    tremor_power_median = tremor_power.median()
+    tremor_power_90th_perc = tremor_power.quantile(config.percentile_tremor_power)
+    
+    # calculate modal tremor power
+    bin_edges = np.linspace(0, 6, 301)
+    kde = gaussian_kde(tremor_power)
+    kde_values = kde(bin_edges)
+    max_index = np.argmax(kde_values)
+    tremor_power_mode = bin_edges[max_index]
+
+    return tremor_power_median, tremor_power_90th_perc, tremor_power_mode
+
+
+def quantify_tremor(df: pd.DataFrame, config: TremorQuantificationConfig, start_time: datetime) -> pd.DataFrame:
+
+    time = [start_time + timedelta(seconds=x) for x in df.time]
+
+    # determine valid days
+    daytime_hours = range(config.daytime_hours_lower_bound,config.daytime_hours_upper_bound)
+    nr_windows_per_day = Counter([dt.date() for dt in time if dt.hour in daytime_hours])
+    valid_days = [day for day, count in nr_windows_per_day.items() 
+                  if count * config.window_length_s / 3600 >= config.valid_day_threshold_hr]
+    df_aggregates = pd.DataFrame({'nr_valid_days': [len(valid_days)]}) # save number of valid days
+
+    # remove windows during non-valid days, non-daytime hours and non-tremor arm movement
+    df_filtered = df.loc[[dt.date() in valid_days for dt in time]]
+    df_filtered = df_filtered.loc[[dt.hour in daytime_hours for dt in time]]
+    df_aggregates['nr_windows_daytime'] = df_filtered.shape[0] # save number of windows during daytime on valid days
+    df_filtered = df_filtered.loc[df_filtered.low_freq_power <= config.movement_threshold]
+    df_aggregates['nr_windows_daytime_rest'] = df_filtered.shape[0] # save number of windows during daytime on valid days without non-tremor arm movement
+
+    # calculate weekly tremor time
+    df_aggregates['tremor_time'] = np.sum(df_filtered['pred_tremor_checked']) / df_filtered.shape[0] * 100
+
+    # calculate weekly tremor power measures
+    tremor_power = df_filtered['tremor_power'][df_filtered['pred_tremor_checked']==1]
+    tremor_power_median, tremor_power_90th_perc, tremor_power_mode = aggregate_tremor_power(tremor_power,config)
+    df_aggregates['tremor_power_median'] = tremor_power_median
+    df_aggregates['tremor_power_mode'] = tremor_power_mode
+    df_aggregates['tremor_power_90th_perc'] = tremor_power_90th_perc
+
+    return df_aggregates
+
+
+def quantify_tremor_io(path_to_feature_input: Union[str, Path], path_to_prediction_input: Union[str, Path], output_path: Union[str, Path], config: TremorQuantificationConfig) -> None:
```

Comment:
Het idee van de _io_ wrapper is dat het alleen de data laadt, een functie draait (de functie naam minus _io_), en de data opslaat. Zou je het zo kunnen omschrijven, en dan wellicht de andere stappen in de functie `quantify_tremor` doen? Wellicht moet de inhoud van `quantify_tremor` dan naar een nieuwe functie, of de andere code naar een nieuwe functie. Your call. ;)

Reply:
Aangepast!